### PR TITLE
chore: use dependabot to manage dependencies

### PR DESCRIPTION
**Pull request by jawnsy**
Use dependabot to manage the dependencies defined in package.json and
GitHub Actions workflows, so that we can proactively update versions.

Outdated versions of third-party dependencies frequently have known
security vulnerabilities with CVEs.

### CODE REVIEW

File: .github/dependabot.yml
Comment (jsjoeio) [inline .github/dependabot.yml:6]:
```suggestion
    interval: "weekly"
```

File: .github/dependabot.yml
Comment (jsjoeio) [inline .github/dependabot.yml:15]:
```suggestion
    interval: weekly
```

File: .github/dependabot.yml
Comment (jsjoeio) [inline .github/dependabot.yml:17]:
```suggestion
  open-pull-requests-limit: 5
```

File: .github/dependabot.yml
Comment (jsjoeio) [inline .github/dependabot.yml:8]:
```suggestion
  open-pull-requests-limit: 5
```

**Review by jsjoeio — CHANGES_REQUESTED** [review]
Thanks for the contribution, @jawnsy! 🎉

In the past when I've used `dependabot` to manage dependencies, it tends to be pretty noisy and can be overwhelming to keep up with. 

Just took a look at the Security tab of our repo and it looks like we can set a lot of this up through there, including dependabot alerts.

image

Do you know the difference between using that vs adding this in like you've done?

**Comment by jsjoeio** [comment]
bump @jawnsy

File: .github/dependabot.yml
Comment (jsjoeio) [reply .github/dependabot.yml:6]: Thanks for the thorough response!

> Just listing out the tradeoffs we're making here as we evaluate this decision. However, since this is an easily reversible decision, regardless of which way we go, I would suggest we make a quick decision and iterate - we can always change the config later.

Sure, I'm all for iterating quickly. Ideally things like this start off as a discussion in an issue rather than a drive-by PR which gives us (maintainers) a chance to ask questions and discuss whether this makes sense for the project before moving to the PR stage. 

> Daily/Weekly schedule interval

I agree with most of the points you're making here but think this fails to recognize an important point: time. Like I said before, it gets noisy and it's hard to keep up with. 

You're coming in here proposing this change, which has value, but are you willing to be the one to maintain it? As you can see from the backlog of issues, we have a lot of other work to be done and merging this comes with a high cost.

> We will be informed about new versions more quickly, which can be important for security releases

I like the sound of that, but do we have evidence to say this is something we need? Probably not?

> Dependabot's PR Workflow vs Console

Thanks for explaining the differences! Sounds like this approach is preferable.

Okay just going to summarize my final thoughts here:
- next time, will you open up an issue so we discuss this before you open a PR?
- we can approve this and merge if you're willing to take ownership on reviewing and merging these PRs

File: .github/dependabot.yml
Comment (jsjoeio) [inline .github/dependabot.yml:4]:
```suggestion
  directory: "/"
  default_assignees:
      - "jawnsy"
```

File: .github/dependabot.yml
Comment (jsjoeio) [inline .github/dependabot.yml:13]:
```suggestion
  directory: "/"
  default_assignees:
    - "jawnsy"
```

**Review by jsjoeio — CHANGES_REQUESTED** [review]

File: .github/dependabot.yml
Comment (jawnsy) [reply .github/dependabot.yml:6]: Sure, I'd be happy to write an issue next time. I didn't bother doing that because I didn't think this change would be objectionable :smile: 

> I like the sound of that, but do we have evidence to say this is something we need? Probably not?

As software maintainers, we're responsible for the supply chain and making sure that we have a clean bill of health with respect to both our code, as well as third-party dependencies that we depend on.

Here's some evidence, the outstanding advisories shown by `yarn audit` run against the latest `main` branch; it's possible that these aren't real issues, but it seems like good hygiene to stay up-to-date, provided that the risk to upgrade isn't significant (and you're in a better position to assess that than I am):

```
[coder@jawnsy-m code-server]$ yarn audit
yarn audit v1.22.10
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ low           │ Prototype Pollution                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ ini                                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >1.3.6                                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ stylelint                                                    │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ stylelint > global-modules > global-prefix > ini             │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1589                        │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Use of a Broken or Risky Cryptographic Algorithm             │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ elliptic                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=6.5.4                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ parcel-bundler                                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ parcel-bundler > node-libs-browser > crypto-browserify >     │
│               │ browserify-sign > elliptic                                   │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1648                        │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Use of a Broken or Risky Cryptographic Algorithm             │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ elliptic                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=6.5.4                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ parcel-bundler                                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ parcel-bundler > node-libs-browser > crypto-browserify >     │
│               │ create-ecdh > elliptic                                       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1648                        │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ high          │ Prototype Pollution in node-forge                            │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ node-forge                                                   │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >= 0.10.0                                                    │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ parcel-bundler                                               │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ parcel-bundler > node-forge                                  │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1561                        │
└───────────────┴──────────────────────────────────────────────────────────────┘
4 vulnerabilities found - Packages audited: 1216
Severity: 1 Low | 2 Moderate | 1 High
Done in 1.44s.
```

File: .github/dependabot.yml
Comment (jawnsy) [reply .github/dependabot.yml:13]: I'd be happy to be the default assignee, but I'm not sure that's actually what you want, because I don't know this project as well as you do, and so wouldn't know whether upgrades are likely to break things. I would need some guidance from you regarding what is needed when deciding whether to approve/merge - what sort of manual testing do I need to do? Or can I just take a quick look at the changes and approve/merge provided that tests are passing?

Comment (jsjoeio) [reply .github/dependabot.yml:13]: > Or can I just take a quick look at the changes and approve/merge provided that tests are passing?

If CI is passing, then it should be okay for you to merge. If it breaks something than I will take responsibility for fixing it and then it will for us to add more tests in order to continue being able to rely on CI and our testing suites.

File: .github/dependabot.yml
Comment (jsjoeio) [reply .github/dependabot.yml:8]: And this one? 5 seems like a great place to start. Like you said, easily reversible so we can increase later if needed.

File: .github/dependabot.yml
Comment (jawnsy) [reply .github/dependabot.yml:13]: lmao, living dangerously, I like it :joy:

Comment (jsjoeio) [reply .github/dependabot.yml:13]: The only way to live 😉

Comment (jsjoeio) [reply .github/dependabot.yml:13]: Okay feel free to make the changes and then we can approve and merge!

Comment (jawnsy) [reply .github/dependabot.yml:13]: Applying this change, but I think you're using the old dependabot syntax (Dependabot Preview), and not the new one, since they were acquired/integrated into GitHub: https://docs.github.com/en/github/administering-a-repository/customizing-dependency-updates

I think we'd want to set me as reviewer, then?

File: .github/dependabot.yml
Comment (jawnsy) [reply .github/dependabot.yml:17]: Looks like 5 is the default, according to the docs, so we can just remove this

> To keep pull requests manageable and easy to review, Dependabot raises a maximum of five pull requests to start bringing dependencies up to the latest version.

From the docs: https://docs.github.com/en/github/administering-a-repository/enabling-and-disabling-version-updates#checking-the-status-of-version-updates

File: .github/dependabot.yml
Comment (jawnsy) [inline .github/dependabot.yml:15]: @jsjoeio Are you OK sticking with daily? I have a preference for daily, but if you feel strongly about it, then I'm happy to go with weekly

Comment (jsjoeio) [reply .github/dependabot.yml:15]: Well they're being assigned and reviewed by you so if you think you can handle daily, all you 😛

File: .github/dependabot.yml
Comment (jsjoeio) [reply .github/dependabot.yml:23]: Oh huh I just pulled that from their docs actually so maybe it's outdated. https://dependabot.com/docs/config-file/#default_assignees

You're right though, we want you as the reviewer (and the assignee, probably). Good catch!

**Comment by code-asher** [comment]
I'm down to own these if that's helpful.

File: .github/dependabot.yml
Comment (jawnsy) [reply .github/dependabot.yml:23]: Yeah, that's for "Dependabot Preview", then they were acquired by GitHub and it became "Dependabot Native" or whatever... at least that's what I was able to gather based on the docs lol

File: .github/dependabot.yml
Comment (jawnsy) [reply .github/dependabot.yml]: Putting this here so we can have a threaded conversation. cc @code-asher 

So, I think this is sorta blocked by #2559. I've learned recently that [Dependabot PR runs in Actions won't receive secrets](https://docs.github.com/en/github/administering-a-repository/keeping-your-actions-up-to-date-with-dependabot#about-dependabot-version-updates-for-actions):

> **Note**: Pull requests for Dependabot version updates will trigger workflow runs with a read-only `GITHUB_TOKEN`. These workflow runs will not be granted access to any secrets.

However, we found in `cdr/coder-cli` that a workaround is to manually re-run the Dependabot PR builds, in which case the run will get secrets since it seems to use the credentials of the person triggering the re-run. This seems like sensible - though non-obvious - behavior.

So I don't know if this is still worth pursuing given the required workaround, or not. What are your thoughts?

File: .github/dependabot.yml
Comment (jsjoeio) [reply .github/dependabot.yml:23]: Ah, that's confusing. Totally missed the yellow box saying that at the top lol

File: .github/dependabot.yml
Comment (code-asher) [reply .github/dependabot.yml]: Personally if it was me reviewing the dependabot PRs I wouldn't mind re-running manually for now until we make the required CI fixes. If anything I'd consider it motivation to get CI fixed. :wink:

**Review by jsjoeio — DISMISSED** [review]
LGTM!

@jawnsy 
good luck han solo gif

File: .github/dependabot.yml
Comment (jawnsy) [inline .github/dependabot.yml:15]: @cdr/code-server-reviewers FYI, I made two minor changes to this since it was last reviewed:

1. Ignore all of the official GitHub Actions actions, we'll need to manage these manually
2. Format dependabot.yml with Prettier as CI needs that

Comment (jsjoeio) [reply .github/dependabot.yml:15]: Sweet! Thanks for the update!

Comment (jawnsy) [reply .github/dependabot.yml:15]: @jsjoeio I'll need your review again, it seems your personal reviews don't count :raised_eyebrow: 

image

Comment (jsjoeio) [reply .github/dependabot.yml:15]: Ah weird! Maybe it's cause you dismissed the stale one? I'll reapprove!
image

**Review by jsjoeio — APPROVED** [review]

File: .github/dependabot.yml
Comment (jawnsy) [reply .github/dependabot.yml:15]: interesting, I wonder if it did that automatically because I added commits, that'd make sense I guess!

Comment (jawnsy) [reply .github/dependabot.yml:15]: Thanks for the approval, excited to see how this goes... :crossed_fingers:

**Pull request merged**
